### PR TITLE
Add support for futex bitset and realtime clock

### PR DESCRIPTION
**Pull request by Sahakait**
Futex syscall now supports `FUTEX_CLOCK_REALTIME` and `FUTEX_{WAKE,WAIT}_BITSET`.

**Changes**:
Thread Queue
image
No function outside of `thread.h` should modify thread queues. Bitset was added to the thread struct. An alternative design was considered with thread queue nodes, such that queue nodes would keep track of threads and their corresponding bitset. This design has the benefit of separating condition logic from the thread struct. However, we would need to `malloc` and `free` queue nodes.
(Below: alternative thread queue design)
image

Real time clock
This is implemented by converting the timeout parameter into a relative timeout.

Bitset support
Most of the change is in `cond.c`. See [futex manpage](https://man7.org/linux/man-pages/man2/futex.2.html) for expected functionality. 

**Testing**:
Added test cases for wait/wake with bitset in `test\futex`.
Added test case for realtime clock in `test\futex`.
Added new test file in `test\thread_queue`

### CODE REVIEW

File: kernel/futex.c [hunk 1]
```diff
@@ -359,19 +375,35 @@ long myst_syscall_futex(
...
+    if (op & FUTEX_CLOCK_REALTIME)
+    {
+        if (!((_op == FUTEX_WAIT_BITSET) ||
+              (_op == FUTEX_WAIT))) // || (op | FUTEX_WAIT_REQUEUE_PI)
+            ERAISE(-EINVAL);
+
+        clock = CLOCK_REALTIME;
+    }
+
+    if (_op == FUTEX_WAIT_BITSET || _op == FUTEX_WAKE_BITSET)
```
[the remark this answers is not in the document]
Comment (Sahakait) [reply kernel/futex.c:392]: Done

File: include/myst/thread.h [hunk 2]
```diff
@@ -81,7 +82,8 @@ struct myst_thread
     uint64_t magic;
 
     /* used by myst_thread_queue_t (condition variables and mutexes) */
-    struct myst_thread* qnext;
+    struct myst_thread* _qnext;
```
[the remark this answers is not in the document]
Comment (Sahakait) [reply include/myst/thread.h]: The change was made to highlight that only the `thread_queue` functions that live inside `thread.h` should access the fields directly. I'm open to suggestions on what the best way to achieve that is.

[the remark this answers is not in the document]
Comment (mingweishih) [reply include/myst/thread.h]: For consistency, I think removing `_` prefix is better. This concept does not make much sense for struct members to me (unless there are precedence).

File: include/myst/thread.h [hunk 3]
```diff
@@ -405,13 +438,66 @@ MYST_INLINE int myst_thread_queue_remove_thread(
     return found ? pos : -1;
 }
 
+/** Find and remove up to `n` threads from the queue such that
+ * <waiting bitset> & <passed in bitset> != 0.
+ * Removed threads are placed in the `matches` queue.
+ */
+MYST_INLINE size_t myst_thread_queue_search_remove_bitset(
```
Comment (mingweishih) [inline include/myst/thread.h:445]: use return type `int` given you're returning -1 on failure

File: kernel/cond.c [hunk 4]
```diff
@@ -141,61 +150,63 @@ int myst_cond_signal_thread(myst_cond_t* c, myst_thread_t* thread)
...
+    myst_thread_queue_t queue = {NULL, NULL};
 
     if (!c)
         return -EINVAL;
 
     myst_spin_lock(&c->lock);
-
-    waiter = myst_thread_queue_pop_front(&c->queue);
-
+    myst_thread_queue_search_remove_bitset(&c->queue, &queue, 1, bitset);
```
Comment (mingweishih) [inline kernel/cond.c:162]: check return

File: kernel/cond.c [hunk 5]
```diff
@@ -141,61 +150,63 @@ int myst_cond_signal_thread(myst_cond_t* c, myst_thread_t* thread)
...
-        /* Select at most n waiters to be woken up */
-        while (i < n && (p = myst_thread_queue_pop_front(&c->queue)))
-        {
-            myst_thread_queue_push_back(&waiters, p);
-            i++;
-        }
-    }
+    myst_spin_lock(&c->lock);
+    /* Select at most n waiters to be woken up */
+    myst_thread_queue_search_remove_bitset(&c->queue, &waiters, n, bitset);
```
Comment (mingweishih) [inline kernel/cond.c:192]: check return

File: include/myst/thread.h [hunk 6]
```diff
@@ -405,13 +438,66 @@ MYST_INLINE int myst_thread_queue_remove_thread(
...
+            num_found++;
+            if (prev != NULL)
+            {
+                prev->_qnext = next;
+                if (next == NULL)
+                    queue->_back = prev;
+            }
+            else
+            {
+                queue->_front = queue->_front->_qnext;
```
Comment (mingweishih) [inline include/myst/thread.h:476]: Isn't `queue->_front->_qnext` the same as `next`? If so, is 477 & 478 the same as 471 & 472?

File: kernel/futex.c [hunk 7]
```diff
@@ -359,19 +375,35 @@ long myst_syscall_futex(
 {
     long ret = 0;
 
-    (void)val3;
+    uint32_t bitset = FUTEX_BITSET_MATCH_ANY;
+    clockid_t clock = CLOCK_MONOTONIC;
+
+    int _op = op & ~(FUTEX_CLOCK_REALTIME | FUTEX_PRIVATE);
```
[the remark this answers is not in the document]
Comment (Sahakait) [reply kernel/futex.c:378]: image
The flags are not designed for this. For example, `op & FUTEX_WAKE` would be true for both `op =  FUTEX_WAKE` and `op = FUTEX_REQUEUE`.

File: include/myst/thread.h (hunk 3 above)
Comment (Sahakait) [reply include/myst/thread.h:463]: Can you elaborate on the reasoning for this? I based this function off of `myst_thread_queue_remove_thread()`.
image

File: include/myst/thread.h [hunk 8]
```diff
@@ -405,13 +438,66 @@ MYST_INLINE int myst_thread_queue_remove_thread(
...
 
+/** Find and remove up to `n` threads from the queue such that
+ * <waiting bitset> & <passed in bitset> != 0.
+ * Removed threads are placed in the `matches` queue.
+ */
+MYST_INLINE size_t myst_thread_queue_search_remove_bitset(
+    myst_thread_queue_t* queue,
+    myst_thread_queue_t* matches,
+    size_t n,
+    uint32_t bitset)
```
[the remark this answers is not in the document]
Comment (Sahakait) [reply include/myst/thread.h:467]: I based the variable name off of the operation name `FUTEX_WAIT_BITSET`. Do you think bitmask would provide more clarity?

File: include/myst/thread.h (hunk 6 above)
Comment (Sahakait) [reply include/myst/thread.h:494]: This handles removing the only node
```
if (queue->_front == NULL)
                    queue->_back = NULL;
```

This handles removing the last node
```
if (next == NULL)
                    queue->_back = prev;

```

Comment (Sahakait) [reply include/myst/thread.h:492]: I simplified the cases.

File: include/myst/cond.h [hunk 9]
```diff
@@ -27,13 +28,20 @@ int myst_cond_timedwait(
     myst_cond_t* c,
     myst_mutex_t* mutex,
     const struct timespec* timeout);
+int myst_cond_timedwait_ops(
```
Comment (mikbras) [inline include/myst/cond.h:31]: Please add blank line between definitions (struct and function).

File: include/myst/cond.h [hunk 10]
```diff
@@ -27,13 +28,20 @@ int myst_cond_timedwait(
...
+    myst_cond_t* c,
+    myst_mutex_t* mutex,
+    const struct timespec* timeout,
+    uint32_t bitset);
 
 int myst_cond_signal(myst_cond_t* c);
+int myst_cond_signal_ops(myst_cond_t* c, uint32_t bitset);
```
Comment (mikbras) [inline include/myst/cond.h:38]: Here too.

File: include/myst/cond.h [hunk 11]
```diff
@@ -27,13 +28,20 @@ int myst_cond_timedwait(
...
 
 int myst_cond_signal_thread(myst_cond_t* c, myst_thread_t* thread);
 
 /* Wake up n waiters */
 int myst_cond_broadcast(myst_cond_t* c, size_t n);
+int myst_cond_broadcast_ops(myst_cond_t* c, size_t n, uint32_t bitset);
```
Comment (mikbras) [inline include/myst/cond.h:44]: Blank line here too.

File: include/myst/futex.h [hunk 12]
```diff
@@ -18,12 +18,21 @@
...
+
 int myst_futex_wait(int* uaddr, int val, const struct timespec* to);
+int myst_futex_wait_ops(
+    int* uaddr,
+    int val,
+    const struct timespec* to,
+    uint32_t bitset);
 
 int myst_futex_wake(int* uaddr, int val);
+int myst_futex_wake_ops(int* uaddr, int val, uint32_t bitset);
```
Comment (mikbras) [inline include/myst/futex.h:36]: Please add blank line before definition to follow existing style.

File: include/myst/futex.h [hunk 13]
```diff
@@ -18,12 +18,21 @@
...
 #define FUTEX_WAIT_BITSET    9
+#define FUTEX_WAKE_BITSET    10
 #define FUTEX_PRIVATE        128
 #define FUTEX_CLOCK_REALTIME 256
 // clang-format on
 
+#define FUTEX_BITSET_MATCH_ANY 0xffffffff
+
 int myst_futex_wait(int* uaddr, int val, const struct timespec* to);
+int myst_futex_wait_ops(
```
Comment (mikbras) [inline include/myst/futex.h:29]: Please add blank line before this. I wish the formatter could do that but unfortunately it cannot.

Comment (mikbras) [inline include/myst/futex.h:29]: I wonder whether ``myst_futex_wait_bitset()`` might be a better name (since that is the only parameter that is different).

File: include/myst/cond.h (hunk 10 above)
Comment (mikbras) [inline include/myst/cond.h:38]: I think I would prefer changing the existing ``myst_cond_signal()`` prototype and adding a ``uint32* bitset`` parameter.

File: include/myst/thread.h [hunk 14]
```diff
@@ -342,50 +344,81 @@ MYST_INLINE bool myst_is_zombied_process(myst_process_t* process)
...
     myst_thread_queue_t* queue)
 {
-    myst_thread_t* thread = queue->front;
+    if (queue)
+        return queue->_front;
+
+    return NULL;
+}
+
+MYST_INLINE myst_thread_t* myst_thread_queue_pop_front_bitset(
```
Comment (mikbras) [inline include/myst/thread.h:393]: The bitset seems unrelated to the thread queue operation (bitset is a member of the thread not the queue).

File: include/myst/thread.h [hunk 15]
```diff
@@ -398,22 +431,22 @@ MYST_INLINE int myst_thread_queue_remove_thread(
...
+                prev->_qnext = t->_qnext;
+                if (t->_qnext == NULL)
+                    queue->_back = prev;
             }
             else
             {
-                queue->front = queue->front->qnext;
-                if (queue->front == NULL)
-                    queue->back = NULL;
+                queue->_front = queue->_front->_qnext;
```
Comment (mikbras) [inline include/myst/thread.h:447]: The prefixing of fields with underscores seems unrelated to this PR and I would ask that it be backed out and discussed separately. I understand the intent is to enforce data encapsulation in the hopes of discouraging use of various fields in certain contexts, but it would be a reversal of the previous authors design decision and I think the discussion could be handled separately.

**Review by mikbras — CHANGES_REQUESTED** [review]

File: include/myst/thread.h (hunk 3 above)
Comment (mingweishih) [reply include/myst/thread.h:463]: The reference you put uses int

File: include/myst/thread.h (hunk 8 above)
[the remark this answers is not in the document]
Comment (mingweishih) [reply include/myst/thread.h:467]: According to https://man7.org/linux/man-pages/man2/futex.2.html, `BITSET` refers to an operation of setting a bit mask.

> FUTEX_WAIT_BITSET (since Linux 2.6.25)
              This operation is like FUTEX_WAIT except that val3 is used
              to provide a 32-bit bit mask to the kernel.

File: include/myst/thread.h (hunk 14 above)
Comment (Sahakait) [reply include/myst/thread.h:393]: the naming of this operation signifies the ability to get the bitset data when popping the thread queue as well. Do you have suggestions on how to handle this?

File: include/myst/thread.h (hunk 3 above)
Comment (Sahakait) [reply include/myst/thread.h:463]: Got it, I initially misread your comment as removing the MYST_INLINE

File: include/myst/thread.h (hunk 8 above)
[the remark this answers is not in the document]
Comment (Sahakait) [reply include/myst/thread.h:467]: I think bitset in this context means "set of bits" rather than "setting bits". For example, C++ has data structure called [bitset](https://www.cplusplus.com/reference/bitset/bitset/). I would prefer to keep the name bitset as it also aligns with the operation name.

File: include/myst/futex.h (hunk 13 above)
Comment (Sahakait) [reply include/myst/futex.h:29]: That was the original name, until I realized that we may have to pass on more parameters for the clock type. I renamed all the operations with additional "optional" arguments to `ops` for consistency.

File: kernel/futex.c [hunk 16]
```diff
@@ -358,20 +374,45 @@ long myst_syscall_futex(
...
 {
     long ret = 0;
+    uint32_t bitset = FUTEX_BITSET_MATCH_ANY;
+    int _op = op & ~(FUTEX_CLOCK_REALTIME | FUTEX_PRIVATE);
 
-    (void)val3;
+    if (op & FUTEX_CLOCK_REALTIME)
+    {
+        if (!((_op == FUTEX_WAIT_BITSET) ||
+              (_op == FUTEX_WAIT))) // || (op | FUTEX_WAIT_REQUEUE_PI)
```
Comment (mikbras) [inline kernel/futex.c:383]: Suggest taking out commented out clause.

**Comment by mikbras** [comment]
Sahakait: This is looking good but I will need more time to review. Will get back to it soon.

File: include/myst/cond.h (hunk 10 above)
Comment (Sahakait) [reply include/myst/cond.h:38]: `pipedev.c` and `eventfddev.c` uses the original function, meaning we would have to add bitset to that context, which doesn't necessarily need to know about it.

File: kernel/futex.c (hunk 16 above)
Comment (Sahakait) [reply kernel/futex.c:383]: I added that there since according to the specs `FUTEX_WAIT_REQUEUE_PI` is allowed to have the `FUTEX_CLOCK_REALTIME` set as well, in case we implement this functionality in the future. If that is more distracting than helpful, I can remove it.

File: include/myst/thread.h (hunk 8 above)
[the remark this answers is not in the document]
Comment (mikbras) [reply include/myst/thread.h:467]: I prefer bitset because it matches the Linux definition.

[the remark this answers is not in the document]
Comment (mikbras) [reply include/myst/thread.h:467]: Also, Linux itself uses the name `bitset` within the kernel in the futex implementation.

File: include/myst/cond.h (hunk 10 above)
Comment (mikbras) [reply include/myst/cond.h:40]: You can pass null into those context.

File: include/myst/futex.h (hunk 13 above)
Comment (mikbras) [reply include/myst/futex.h:30]: But currently, the only additional parameter is ``bitset``. I think just adding additional (possibly null) parameters to ``myst_futex_wait`` generates fewer changes.

File: include/myst/thread.h [hunk 17]
```diff
@@ -5,6 +5,7 @@
 #define _MYST_THREAD_H
```
Comment (mikbras) [inline include/myst/thread.h:5]: Please don't use this technique to work around circular dependencies. I think there is a better way to resolve.

File: include/myst/thread.h [hunk 18]
```diff
@@ -208,6 +209,7 @@ struct myst_thread
 
     /* used by myst_thread_queue_t (condition variables and mutexes) */
     struct myst_thread* qnext;
+    uint64_t bitset;
```
Comment (mikbras) [inline include/myst/thread.h:212]: Should be `uint32_t since `bitset` is `uint32_t` elsewhere and the mask itself is 32-bits.

File: include/myst/thread.h [hunk 19]
```diff
@@ -370,8 +374,25 @@ MYST_INLINE void myst_thread_queue_push_back(
...
 
-MYST_INLINE myst_thread_t* myst_thread_queue_pop_front(
+MYST_INLINE void myst_thread_queue_push_back(
+    myst_thread_queue_t* queue,
+    myst_thread_t* thread)
+{
+    myst_thread_queue_push_back_bitset(queue, thread, FUTEX_BITSET_MATCH_ANY);
+}
+
+MYST_INLINE myst_thread_t* myst_thread_queue_get_front(
```
Comment (mikbras) [inline include/myst/thread.h:384]: I don't see the need for this function. Also it introduces a branch that was not in the original code. It is used in contexts where the queue is known to be non-null.

File: kernel/cond.c [hunk 20]
```diff
@@ -34,7 +34,7 @@ int myst_cond_destroy(myst_cond_t* c)
     myst_spin_lock(&c->lock);
 
     /* Fail if queue is not empty */
-    if (c->queue.front)
+    if (!myst_thread_queue_empty(&c->queue))
```
Comment (mikbras) [inline kernel/cond.c:37]: I would like to remove this accessor since the only consumer of this field is the condition itself.

File: kernel/futex.c [hunk 21]
```diff
@@ -269,6 +280,11 @@ int myst_futex_wake(int* uaddr, int val)
     return ret;
 }
 
+int myst_futex_wake(int* uaddr, int val)
+{
+    return myst_futex_wake_ops(uaddr, val, FUTEX_BITSET_MATCH_ANY);
```
Comment (mikbras) [inline kernel/futex.c:285]: Does this function now require checking of futex-bitsets? If so, it modifies its original behavior for the non-bitset case, which could introduce instability. I would prefer that existing code paths remain exactly as before.

File: kernel/futex.c [hunk 22]
```diff
@@ -358,20 +374,45 @@ long myst_syscall_futex(
     int val3)
 {
     long ret = 0;
+    uint32_t bitset = FUTEX_BITSET_MATCH_ANY;
+    int _op = op & ~(FUTEX_CLOCK_REALTIME | FUTEX_PRIVATE);
```
Comment (mikbras) [inline kernel/futex.c:378]: we use underscore prefixes to denote static definitions.

File: kernel/mutex.c [hunk 23]
```diff
@@ -8,6 +8,7 @@
 #include <stdbool.h>
 #include <string.h>
 
+#include <myst/cond.h>
```
Comment (mikbras) [inline kernel/mutex.c:11]: Including cond.h here introduces a circular inclusion dependency (cond.h also includes mutex.h). This was made necessary be the definition of the `myst_thread_queue_empty()`` accessors which I would remove.

File: kernel/mutex.c [hunk 24]
```diff
@@ -42,7 +43,7 @@ int __myst_mutex_trylock(myst_mutex_t* m, myst_thread_t* self)
     if (m->owner == NULL)
     {
         /* If the waiters queue is empty */
-        if (m->queue.front == NULL)
```
Comment (mikbras) [inline kernel/mutex.c:45]: All changes to mutex.h are unecessary and were introduced by the accessors for condition. Mutexes and conditions are tightly coupled intentionally and mutex should not be considered a consumer of condition.

**Comment by mikbras** [comment]
Overall the spirit of the change looks good, but I have the following concerns.
- The change may possibly modify existing behavior when bit sets are not in use. I need to spend more time looking at this.
- There are preferential changes that complicate the PR and make it hard to review. The use of accessors for example (between mutex and condition) creates a lot of churn and makes it hard to really see the essential changes. Also the accessors have required header inclusion cycles.

**Comment by mikbras** [comment]
Sakakait: any idea why this fails the pipeline. It looks like the Java tests fails and crashes before running other tests.

File: include/myst/thread.h (hunk 17 above)
Comment (mikbras) [reply include/myst/thread.h:5]: Please disregard. I am concerned about circular inclusions elsewhere though.

**Comment by mikbras** [comment]
Hi Sahakai:

I checked out your branch and spent some time examining and testing. The futex tests fails often under stress. I run the following script on my machine (you will have to install sysbench on your machine).

Here is my ``stress`` script:

```C++
#!/bin/bash
while true
do
        sysbench --test=cpu --cpu-max-prime=10000000 --threads=16 run
done
```

**Comment by Sahakait** [comment]
> Sakakait: any idea why this fails the pipeline. It looks like the Java tests fails and crashes before running other tests.

I just rebased and I see additional failures, I suspect this may have to do with merging this PR with other changes in `cond.c`, will need to investigate further.

**Comment by Sahakait** [comment]
> Hi Sahakait:
> 
> I checked out your branch and spent some time examining and testing. The futex tests fails often under stress. I run the following script on my machine (you will have to install `sysbench` on your machine).
> 
> ```c++
> #!/bin/bash
> while true
> do
>         sysbench --test=cpu --cpu-max-prime=10000000 --threads=16 run
> done
> ```
> 
> Also, `tests/hello_world/java` hangs on my machine.

This is likely due to the fact that the wait tests uses a timer to make sure the wait duration is within a certain threshold of the specified wait time.

File: kernel/mutex.c (hunk 24 above)
Comment (Sahakait) [reply kernel/mutex.c:45]: `kernel/mutex.c` has been reverted.

File: kernel/mutex.c (hunk 23 above)
Comment (Sahakait) [reply kernel/mutex.c:11]: `kernel/mutex.c` has been reverted.

File: kernel/futex.c (hunk 22 above)
Comment (Sahakait) [reply kernel/futex.c:378]: `_op` renamed.

File: kernel/cond.c (hunk 20 above)
Comment (Sahakait) [reply kernel/cond.c:37]: Reverted.

**Comment by Sahakait** [comment]
Resubmitted as a separate PR.
